File: finverse_api/app/services/recurring_transaction_service.py
"""
Service for managing recurring transactions in FinVerse API
"""

from typing import List, Optional
from datetime import date, datetime, timedelta
from fastapi import HTTPException, status
from sqlalchemy.orm import Session
from sqlalchemy.exc import SQLAlchemyError

from app.models.recurring_transaction import RecurringTransaction
from app.schemas.recurring_transaction import RecurringTransactionCreate, RecurringTransactionUpdate
from app.schemas.recurring_transaction import FrequencyType


class RecurringTransactionService:
    """Service for managing recurring transactions"""
    
    @staticmethod
    def get_recurring_transactions(db: Session, user_id: int) -> List[RecurringTransaction]:
        """Get all recurring transactions for a user"""
        return db.query(RecurringTransaction).filter(
            RecurringTransaction.user_id == user_id
        ).order_by(RecurringTransaction.next_occurrence).all()
    
    @staticmethod
    def get_recurring_transaction(db: Session, transaction_id: int, user_id: int) -> RecurringTransaction:
        """Get a specific recurring transaction by ID"""
        transaction = db.query(RecurringTransaction).filter(
            RecurringTransaction.id == transaction_id,
            RecurringTransaction.user_id == user_id
        ).first()
        
        if not transaction:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail=f"Recurring transaction with ID {transaction_id} not found"
            )
        
        return transaction
    
    @staticmethod
    def create_recurring_transaction(
        db: Session, transaction_data: RecurringTransactionCreate, user_id: int
    ) -> RecurringTransaction:
        """Create a new recurring transaction"""
        try:
            # Calculate next occurrence based on start date and frequency
            next_occurrence = RecurringTransactionService._calculate_next_occurrence(
                transaction_data.start_date,
                transaction_data.frequency_type,
                transaction_data.frequency_value
            )
            
            transaction = RecurringTransaction(
                user_id=user_id,
                category_id=transaction_data.category_id,
                wallet_id=transaction_data.wallet_id,
                amount=transaction_data.amount,
                transaction_type=transaction_data.transaction_type,
                description=transaction_data.description,
                frequency_type=transaction_data.frequency_type,
                frequency_value=transaction_data.frequency_value,
                start_date=transaction_data.start_date,
                end_date=transaction_data.end_date,
                next_occurrence=next_occurrence,
                is_active=transaction_data.is_active
            )
            
            db.add(transaction)
            db.commit()
            db.refresh(transaction)
            return transaction
            
        except SQLAlchemyError as e:
            db.rollback()
            raise HTTPException(
                status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
                detail=f"Database error: {str(e)}"
            )
    
    @staticmethod
    def update_recurring_transaction(
        db: Session, 
        transaction_id: int, 
        transaction_data: RecurringTransactionUpdate, 
        user_id: int
    ) -> RecurringTransaction:
        """Update an existing recurring transaction"""
        transaction = RecurringTransactionService.get_recurring_transaction(db, transaction_id, user_id)
        
        try:
            update_data = transaction_data.dict(exclude_unset=True)
            
            # If frequency or start date is updated, recalculate next occurrence
            recalculate_next = False
            if "frequency_type" in update_data or "frequency_value" in update_data or "start_date" in update_data:
                recalculate_next = True
                
            # Update transaction fields
            for key, value in update_data.items():
                setattr(transaction, key, value)
            
            # Recalculate next occurrence if needed
            if recalculate_next:
                transaction.next_occurrence = RecurringTransactionService._calculate_next_occurrence(
                    transaction.start_date,
                    transaction.frequency_type,
                    transaction.frequency_value,
                    from_date=datetime.now().date()
                )
            
            db.commit()
            db.refresh(transaction)
            return transaction
            
        except SQLAlchemyError as e:
            db.rollback()
            raise HTTPException(
                status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
                detail=f"Database error: {str(e)}"
            )
    
    @staticmethod
    def delete_recurring_transaction(db: Session, transaction_id: int, user_id: int) -> bool:
        """Delete a recurring transaction"""
        transaction = RecurringTransactionService.get_recurring_transaction(db, transaction_id, user_id)
        
        try:
            db.delete(transaction)
            db.commit()
            return True
        except SQLAlchemyError as e:
            db.rollback()
            raise HTTPException(
                status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
                detail=f"Database error: {str(e)}"
            )
    
    @staticmethod
    def get_due_recurring_transactions(db: Session, current_date: date = None) -> List[RecurringTransaction]:
        """Get all recurring transactions that are due for processing"""
        if current_date is None:
            current_date = datetime.now().date()
        
        return db.query(RecurringTransaction).filter(
            RecurringTransaction.is_active == True,
            RecurringTransaction.next_occurrence <= current_date,
            (RecurringTransaction.end_date.is_(None) | (RecurringTransaction.end_date >= current_date))
        ).all()
    
    @staticmethod
    def update_next_occurrence(db: Session, transaction_id: int) -> RecurringTransaction:
        """Update the next occurrence date after a transaction is processed"""
        transaction = db.query(RecurringTransaction).filter(
            RecurringTransaction.id == transaction_id
        ).first()
        
        if not transaction:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail=f"Recurring transaction with ID {transaction_id} not found"
            )
        
        try:
            current_date = datetime.now().date()
            
            # If end date exists and we've reached it, deactivate the recurring transaction
            if transaction.end_date and transaction.end_date <= current_date:
                transaction.is_active = False
                db.commit()
                return transaction
            
            # Calculate next occurrence date
            next_date = RecurringTransactionService._calculate_next_occurrence(
                transaction.start_date,
                transaction.frequency_type,
                transaction.frequency_value,
                from_date=current_date
            )
            
            transaction.next_occurrence = next_date
            db.commit()
            db.refresh(transaction)
            return transaction
            
        except SQLAlchemyError as e:
            db.rollback()
            raise HTTPException(
                status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
                detail=f"Database error: {str(e)}"
            )
    
    @staticmethod
    def _calculate_next_occurrence(
        start_date: date,
        frequency_type: int,
        frequency_value: int,
        from_date: date = None
    ) -> date:
        """
        Calculate the next occurrence date based on frequency settings
        
        Args:
            start_date: The start date of the recurring transaction
            frequency_type: The type of frequency (1=daily, 2=weekly, 3=monthly, 4=yearly)
            frequency_value: The value for the frequency
            from_date: The date to calculate from (defaults to start_date)
        
        Returns:
            The next occurrence date
        """
        if from_date is None:
            from_date = start_date
        
        # If from_date is before start_date, use start_date
        if from_date < start_date:
            from_date = start_date
        
        # Daily frequency
        if frequency_type == FrequencyType.DAILY:
            return from_date + timedelta(days=1)
        
        # Weekly frequency
        elif frequency_type == FrequencyType.WEEKLY:
            # frequency_value is day of week (0-6, Monday=0, Sunday=6)
            days_ahead = frequency_value - from_date.weekday()
            if days_ahead <= 0:  # Target day already happened this week
                days_ahead += 7
            return from_date + timedelta(days=days_ahead)
        
        # Monthly frequency
        elif frequency_type == FrequencyType.MONTHLY:
            # frequency_value is day of month (1-31)
            next_month = from_date.replace(day=1)
            if from_date.day >= frequency_value:
                # Move to next month if we're already past the target day
                if from_date.month == 12:
                    next_month = next_month.replace(year=from_date.year + 1, month=1)
                else:
                    next_month = next_month.replace(month=from_date.month + 1)
            
            # Calculate the actual day, adjusting for shorter months
            import calendar
            last_day = calendar.monthrange(next_month.year, next_month.month)[1]
            target_day = min(frequency_value, last_day)
            
            return next_month.replace(day=target_day)
        
        # Yearly frequency
        elif frequency_type == FrequencyType.YEARLY:
            # frequency_value is day of year (1-366)
            from dateutil.relativedelta import relativedelta
            
            current_year_date = RecurringTransactionService._get_day_of_year(from_date.year, frequency_value)
            
            # If we've already passed this date in the current year, move to next year
            if current_year_date < from_date:
                return RecurringTransactionService._get_day_of_year(from_date.year + 1, frequency_value)
            return current_year_date
        
        # Default fallback
        return from_date
    
    @staticmethod
    def _get_day_of_year(year: int, day_of_year: int) -> date:
        """Helper method to get a date from a day of year value"""
        import datetime as dt
        
        # Handle leap years
        if day_of_year > 366:
            day_of_year = 366
        
        try:
            return dt.date(year, 1, 1) + dt.timedelta(days=day_of_year - 1)
        except ValueError:
            # Handle Feb 29 in non-leap years
            return dt.date(year, 12, 31) 

File: finverse_ui/src/services/recurringTransactionService.ts
import { AxiosError } from 'axios';
import api from './api';
import { handleErrorResponse } from '../utils/importFixes';

// Types for recurring transactions
export interface RecurringTransaction {
  id: number;
  user_id: number;
  category_id: number;
  wallet_id: number;
  amount: number;
  transaction_type: number; // 0=expense, 1=income
  description?: string;
  frequency_type: number; // 1=daily, 2=weekly, 3=monthly, 4=yearly
  frequency_value: number;
  start_date: string; // ISO format
  end_date?: string; // ISO format
  next_occurrence: string; // ISO format
  is_active: boolean;
  created_at: string; // ISO format
  updated_at: string; // ISO format
}

export interface RecurringTransactionCreate {
  category_id: number;
  wallet_id: number;
  amount: number;
  transaction_type: number;
  description?: string;
  frequency_type: number;
  frequency_value: number;
  start_date: string;
  end_date?: string;
  is_active: boolean;
}

export interface RecurringTransactionUpdate {
  category_id?: number;
  wallet_id?: number;
  amount?: number;
  transaction_type?: number;
  description?: string;
  frequency_type?: number;
  frequency_value?: number;
  start_date?: string;
  end_date?: string;
  is_active?: boolean;
}

// Error response interface
interface ApiErrorResponse {
  detail?: string;
  message?: string;
}

// Enum-like interfaces for frequency and transaction types
export const FrequencyType = {
  DAILY: 1,
  WEEKLY: 2,
  MONTHLY: 3,
  YEARLY: 4,
} as const;

export const TransactionType = {
  EXPENSE: 0,
  INCOME: 1,
} as const;

// Human-readable labels for the frequency types
export const FrequencyTypeLabels = {
  [FrequencyType.DAILY]: 'Daily',
  [FrequencyType.WEEKLY]: 'Weekly',
  [FrequencyType.MONTHLY]: 'Monthly',
  [FrequencyType.YEARLY]: 'Yearly',
};

// Human-readable labels for the transaction types
export const TransactionTypeLabels = {
  [TransactionType.EXPENSE]: 'Expense',
  [TransactionType.INCOME]: 'Income',
};

// Helper function to format frequency value based on type
export const formatFrequency = (type: number, value: number): string => {
  const days = ['Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday', 'Sunday'];
  
  switch (type) {
    case FrequencyType.DAILY:
      return 'Every day';
    case FrequencyType.WEEKLY:
      return `Every ${days[value]}`;
    case FrequencyType.MONTHLY:
      return `Day ${value} of every month`;
    case FrequencyType.YEARLY: {
      // Convert day of year to date (without year)
      const date = new Date(new Date().getFullYear(), 0, value);
      return `Every ${date.toLocaleDateString('en-US', { month: 'long', day: 'numeric' })}`;
    }
    default:
      return '';
  }
};

// Helper to extract error message from API error
const getErrorDetail = (responseData: unknown): string => {
  if (responseData && typeof responseData === 'object') {
    const errorData = responseData as ApiErrorResponse;
    return errorData.detail || errorData.message || 'Please check your input and try again.';
  }
  return 'Please check your input and try again.';
};

// Service for recurring transactions
export const recurringTransactionService = {
  // Get all recurring transactions for the current user
  getRecurringTransactions: async (): Promise<RecurringTransaction[]> => {
    try {
      const response = await api.get('/recurring');
      return response.data.recurring_transactions;
    } catch (error) {
      const err = error as AxiosError;
      console.error('Error fetching recurring transactions:', err);
      throw err;
    }
  },

  // Get a specific recurring transaction by ID
  getRecurringTransaction: async (id: number): Promise<RecurringTransaction> => {
    try {
      const response = await api.get(`/recurring/${id}`);
      return response.data;
    } catch (error) {
      const err = error as AxiosError;
      console.error(`Error fetching recurring transaction #${id}:`, err);
      throw err;
    }
  },

  // Create a new recurring transaction
  createRecurringTransaction: async (
    transaction: RecurringTransactionCreate
  ): Promise<RecurringTransaction> => {
    try {
      console.log('API Request body:', transaction);
      const response = await api.post('/recurring', transaction);
      return response.data;
    } catch (error) {
      const err = error as AxiosError;
      console.error('Error creating recurring transaction:', err);
      
      // Handle validation errors (422)
      if (err.response?.status === 422) {
        console.error('Validation error details:', err.response.data);
        throw new Error(`Invalid data: ${getErrorDetail(err.response.data)}`);
      } else if (err.response?.status === 500) {
        throw new Error('Server error occurred while creating recurring transaction. Please try again or contact support.');
      } else if (err.response?.status === 400) {
        throw new Error(`Invalid data: ${getErrorDetail(err.response.data)}`);
      }
      throw err;
    }
  },

  // Update an existing recurring transaction
  updateRecurringTransaction: async (
    id: number,
    transaction: RecurringTransactionUpdate
  ): Promise<RecurringTransaction> => {
    try {
      const response = await api.put(`/recurring/${id}`, transaction);
      return response.data;
    } catch (error) {
      const err = error as AxiosError;
      console.error(`Error updating recurring transaction #${id}:`, err);
      // Add better error handling with user-friendly messages
      if (err.response?.status === 500) {
        throw new Error('Server error occurred while updating recurring transaction. Please try again or contact support.');
      } else if (err.response?.status === 400) {
        throw new Error(`Invalid data: ${getErrorDetail(err.response.data)}`);
      } else if (err.response?.status === 404) {
        throw new Error('Transaction not found. It may have been deleted.');
      }
      throw err;
    }
  },

  // Delete a recurring transaction
  deleteRecurringTransaction: async (id: number): Promise<boolean> => {
    try {
      const response = await api.delete(`/recurring/${id}`);
      return response.data.success;
    } catch (error) {
      const err = error as AxiosError;
      console.error(`Error deleting recurring transaction #${id}:`, err);
      // Add better error handling with user-friendly messages
      if (err.response?.status === 500) {
        throw new Error('Server error occurred while deleting recurring transaction. Please try again or contact support.');
      } else if (err.response?.status === 404) {
        throw new Error('Transaction not found. It may have been deleted already.');
      }
      throw err;
    }
  },

  // Process a recurring transaction (update next occurrence)
  processRecurringTransaction: async (id: number): Promise<RecurringTransaction> => {
    try {
      const response = await api.post(`/recurring/${id}/process`);
      return response.data;
    } catch (error) {
      const err = error as AxiosError;
      console.error(`Error processing recurring transaction #${id}:`, err);
      // Add better error handling with user-friendly messages
      if (err.response?.status === 500) {
        throw new Error('Server error occurred while processing recurring transaction. Please try again or contact support.');
      } else if (err.response?.status === 404) {
        throw new Error('Transaction not found. It may have been deleted.');
      }
      throw err;
    }
  }
};